zerver/migrations/0209_subscription_notifications_data_migration.py

```diff
@@ -0,0 +1,65 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.18 on 2019-02-13 20:13
+from __future__ import unicode_literals
+
+from django.db import migrations, models
+from django.db.backends.postgresql_psycopg2.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+RECIPIENT_STREAM = 2
+DEFAULT_VALUES = {
+    'desktop_notifications': {
+        'default_for_streams': lambda user_profile: user_profile.enable_stream_desktop_notifications,
+        'default_for_pms': lambda user_profile: user_profile.enable_desktop_notifications,
+    },
+    'audible_notifications': {
+        'default_for_streams': lambda user_profile: user_profile.enable_stream_sounds,
+        'default_for_pms': lambda user_profile: user_profile.enable_sounds,
+    },
+    'push_notifications': {
+        'default_for_streams': lambda user_profile: user_profile.enable_stream_push_notifications,
+        'default_for_pms': lambda user_profile: (user_profile.enable_online_push_notifications or
+                                                 user_profile.enable_offline_push_notifications),
+    },
+    'email_notifications': {
+        'default_for_streams': lambda user_profile: user_profile.enable_stream_email_notifications,
+        'default_for_pms': lambda user_profile: user_profile.enable_offline_email_notifications,
+    },
+}
+
+def update_notification_settings(apps: StateApps, schema_editor: DatabaseSchemaEditor) -> None:
+    Subscription = apps.get_model('zerver', 'Subscription')
+
+    for sub in Subscription.objects.select_related('user_profile', 'recipient'):
+        for setting_name, setting in DEFAULT_VALUES.items():
+            if sub.recipient.type == RECIPIENT_STREAM:
+                default_value = setting['default_for_streams'](sub.user_profile)
+            else:
+                default_value = setting['default_for_pms'](sub.user_profile)
+            if getattr(sub, setting_name) == default_value:
+                setattr(sub, setting_name, None)
+        sub.save()
+
```

Comment:
This migration is going to be a performance problem for Zulip Cloud, unfortunately.  Even for chat.zulip.org, there are 300K Subscription objects (!).  I think there's few things we can do to optimize this:
* Use `atomic=False` on the migration; this avoids taking big locks on the database.
* Do one query per user/attribute pair, not one per subscription/attribute pair.  The default is the same for every stream for a given user, and we can do these as bulk updates:
```
filter_kwargs = {setting_name: default_value}
update_kwargs = {setting_name: None}
Subscription.objects.filter(user_profile=user_profile, **kwargs).update(**update_kwargs)
```